File: Algorithm_problem_solving/Baek-joon/1021/1021.py
import sys

input = sys.stdin.readline

# 덱을 위한 노드 클래스
class Node:
    def __init__(self, item):
        self.item = item
        self.l_node = None
        self.r_node = None

# 덱 클래스
class Deque:
    def __init__(self):
        self.cnt = 0
        self.head = None
        self.tail = None

    def size(self):
        return self.cnt

    def empty(self):
        return True if self.cnt == 0 else False

    def front(self):
        if self.empty():
            return -1
        return self.head.item

    def back(self):
        if self.empty():
            return -1
        return self.tail.item

    def push_front(self, item):
        new_node = Node(item)
        if self.head:
            new_node.r_node = self.head
            self.head.l_node = new_node
        else:
            self.tail = new_node
        self.head = new_node
        self.cnt += 1

    def push_back(self, item):
        new_node = Node(item)
        if self.head:
            new_node.l_node = self.tail
            self.tail.r_node = new_node
        else:
            self.head = new_node
        self.tail = new_node
        self.cnt += 1

    def pop_front(self):
        if self.empty():
            return -1
        output = self.front()
        temp = self.head
        self.head = self.head.r_node
        if self.head:
            self.head.l_node = None
        else:
            self.tail = None
        del temp
        self.cnt -= 1
        return output

    def pop_back(self):
        if self.empty():
            return -1
        output = self.back()
        temp = self.tail
        self.tail = self.tail.l_node
        if self.tail:
            self.tail.r_node = None
        else:
            self.head = None
        del temp
        self.cnt -= 1
        return output

    def printall(self):
        node = self.head
        while node:
            print(node.item)
            node = node.r_node


if __name__ == "__main__":
    N, M = map(int, input().strip().split())
    wants = list(map(int, input().strip().split()))
    cnt = 0
    seqeunce = 0
    deque = Deque()

    # N까지의 값을 덱에 집어 넣는다.
    for i in range(1, N+1):
        deque.push_back(i)

    for want in wants:
        # 원하는 수가 덱에 앞쪽에 있는지 뒤쪽에 있는지 확인하는 변수
        pos = (want - 1 + seqeunce) % deque.cnt
        # 앞쪽이라면 앞부터 pop한다
        if pos < deque.cnt//2:
            while True:
                out = deque.pop_front()
                if out == want:
                    # deque.push_back(out)
                    break
                else:
                    seqeunce -= 1
                    deque.push_back(out)
                    cnt += 1
        # 뒤라면 뒤에서 팝한다.
        else:
           while True:
                out = deque.pop_back()
                if out == want:
                    # deque.push_front(out)
                    break
                else:
                    seqeunce += 1
                    deque.push_front(out)
                    cnt += 1

    print(cnt)

File: Algorithm_problem_solving/Baek-joon/1021/1021.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Deque struct {
	arr []int
}

func NewDeque() *Deque {
	deque := Deque{}
	deque.arr = []int{}
	return &deque
}

func (d Deque) front() int {
	// get deque's first item
	return d.arr[0]
}

func (d Deque) back() int {
	// get deque's last item
	return d.arr[d.size()-1]
}

func (d Deque) size() int {
	// get deque's size
	return len(d.arr)

}

func (d *Deque) popFront() (int, bool) {
	// pop deque's first item
	if d.size() <= 0 {
		return 0, false
	}
	var output int
	output, d.arr = d.arr[0], d.arr[1:]
	return output, true
}

func (d *Deque) popBack() (int, bool) {
	// pop deque's last item
	if d.size() <= 0 {
		return 0, false
	}
	var output int
	d.arr, output = d.arr[:d.size()-2], d.size()-1
	return output, true
}

func (d *Deque) insertFront(n int) {
	// insert n into deque's first position
	d.arr = append([]int{n}, d.arr...)
}

func (d *Deque) insertBack(n int) {
	// insert n into deque's last position
	d.arr = append(d.arr, n)
}

func (d Deque) findIndex(n int) int {
	// find n's postion in deque
	output := -1
	for idx, item := range d.arr {
		if item == n {
			return idx
		}
	}
	return output
}

func (d *Deque) rotateLeft(n int) {
	// roatate queue items into left
	var left, right []int
	left, right = d.arr[:n], d.arr[n:]
	d.arr = append(right, left...)
}

func (d *Deque) rotateRight(n int) {
	// roatate queue items into right
	var left, right []int
	left, right = d.arr[:d.size()-n], d.arr[d.size()-n:]
	d.arr = append(right, left...)
}

func main() {
	var n int
	var arr string
	var numList []string

	deque := NewDeque()
	reader := bufio.NewReader(os.Stdin)

	arr, _ = reader.ReadString('\n')
	arr = strings.Trim(arr, "\n")
	numList = strings.Split(arr, " ")

	n, _ = strconv.Atoi(numList[0])

	arr, _ = reader.ReadString('\n')
	arr = strings.Trim(arr, "\n")
	numList = strings.Split(arr, " ")

	for i := 0; i < n; i++ {
		deque.insertBack(i + 1)
	}

	answer := 0
	for _, item := range numList {
		num, _ := strconv.Atoi(item)
		index := deque.findIndex(num)

		if index <= deque.size()-index {
			deque.rotateLeft(index)
			answer += index
		} else {
			deque.rotateRight(deque.size() - index)
			answer += deque.size() - index
		}
		deque.popFront()
	}
	fmt.Println(answer)
}
